extractor/src/main/java/org/schabi/newpipe/extractor/stream/DeliveryFormat.java

```diff
@@ -0,0 +1,134 @@
+package org.schabi.newpipe.extractor.stream;
+
+import java.io.Serializable;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+
+/**
+ * A class that is used to represent the way that a streaming service deliver their streams.
+ */
+public abstract class DeliveryFormat implements Serializable {
+
+    public static Direct direct(@NonNull String url) {
+        return new Direct(url);
+    }
+
+    public static HLS hls(@NonNull String url) {
+        return new HLS(url);
+    }
+
+    public static ManualDASH manualDASH(@NonNull String baseUrl,
+                                        @NonNull String manualDashManifest) {
+        return new ManualDASH(baseUrl, manualDashManifest);
+    }
+
+    /**
+     * Used when a service offer a direct link to their streams.
+     */
+    public static class Direct extends DeliveryFormat {
+        private final String url;
+
+        private Direct(String url) {
+            this.url = url;
+        }
+
+        public String getUrl() {
+            return url;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+
+            Direct direct = (Direct) o;
+
+            //noinspection EqualsReplaceableByObjectsCall
+            return url != null ? url.equals(direct.url) : direct.url == null;
+        }
+
+        @Override
+        public int hashCode() {
+            return url != null ? url.hashCode() : 0;
+        }
+    }
+
+    /**
+     * Used when a service uses HLS playlists for delivering their content.
+     */
+    public static class HLS extends DeliveryFormat {
+        private final String url;
+
+        private HLS(String url) {
+            this.url = url;
+        }
+
+        public String getUrl() {
+            return url;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+
+            HLS hls = (HLS) o;
+
+            //noinspection EqualsReplaceableByObjectsCall
+            return url != null ? url.equals(hls.url) : hls.url == null;
+        }
+
+        @Override
+        public int hashCode() {
+            return url != null ? url.hashCode() : 0;
```

Comment:
Same for `Objects.hashCode()`.

```suggestion
            return Objects.hashCode(url);
```